### PR TITLE
Add codecov step in CI and add badges to README

### PR DESCRIPTION
**Pull request by yuxincs**
This PR adds a codecov step in CI configs and add the corresponding badges to README.

We have also added a config file for codecov to fail PRs if the coverage decreases (experimental, we can revert / disable it if it gets too annoying in the future)

### CODE REVIEW

File: .codecov.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  range: 80..100
+  round: down
+  precision: 2
+
+  status:
+    project:                   # measuring the overall project coverage
+      default:                 # context, you can create multiple ones with custom titles
+        enabled: yes           # must be yes|true to enable this status
+        target: auto           # specify the target coverage for each commit status
+                               #   option: "auto" (must increase from parent commit or pull request base)
```
Comment (lazaroclapp) [inline .codecov.yaml:11]:
```suggestion
                               #   option: "auto" (must not decrease from parent commit or pull request base)
```

Right?

**Review by lazaroclapp — APPROVED** [review]
🚀

File: .codecov.yaml (hunk 1 above)
Comment (yuxincs) [reply .codecov.yaml]: Right, actually I found out that we can do `auto` with a `threshold` https://docs.codecov.com/docs/commit-status and I have set this to be 5%, meaning that the coverage cannot drop more than 5% (to also allow some flakiness). 

Let me know if you think we should be stricter about this :)

Comment (lazaroclapp) [reply .codecov.yaml]: SGTM!

**Review by lazaroclapp — APPROVED** [review]
Actually, can we make the threshold drop 1-2%. Given the size of the codebase, coverage going down by 5% can easily be a pretty huge PR with no tests. No need for me to re-stamp, though.

**Pull request merged**